### PR TITLE
Bulk momentum

### PR DESCRIPTION
**Pull request by atmyers**

### CODE REVIEW

File: Source/Initialization/InjectorMomentum.H [hunk 1]
```diff
@@ -108,6 +125,16 @@ struct InjectorMomentumBoltzmann
         return amrex::XDim3 {u[0],u[1],u[2]};
     }
 
+    AMREX_GPU_HOST_DEVICE
+    amrex::XDim3
+    getBulkMomentum (amrex::Real x, amrex::Real y, amrex::Real z) const noexcept
+    {
+        amrex::Real u[3];
+        for (int idim = 0; idim < 3; ++idim) u[idim] = 0.0;
+        u[dir] = beta*PhysConst::c;
```
[the remark this answers is not in the document]
Comment (MaxThevenet) [reply Source/Initialization/InjectorMomentum.H]: From the utilization we do below in `PhysicalParticleContainer::AddPlasma`
```c++
u = inj_mom->getBulkMomentum(x, y, 0.);
Real gamma = std::sqrt( 1.+(u.x*u.x+u.y*u.y+u.z*u.z) );
```
or
```c++
u.z = gamma_boost * ( u.z -beta_boost*gamma_lab );
```
I would say we should return `gamma*beta`.

File: Source/Particles/PhysicalParticleContainer.cpp [hunk 2]
```diff
@@ -672,12 +672,21 @@ PhysicalParticleContainer::AddPlasma (int lev, RealBox part_realbox)
...
                 // xmin, xmax, ymin, ymax, zmin, zmax, go to
                 // the next generated particle.
-                if (!inj_pos->insideBounds(xb, yb, z)) {
+
+                // include ballistic correction for plasma species with bulk motion
+                // note that this assumes that momentum profile is independent of z
+                u = inj_mom->getBulkMomentum(x, y, 0.);
+                Real gamma = std::sqrt( 1.+(u.x*u.x+u.y*u.y+u.z*u.z) );
+                Real betaz = u.z/gamma;
+                Real z0 = z - PhysConst::c*t*betaz;
```
Comment (MaxThevenet) [inline Source/Particles/PhysicalParticleContainer.cpp]: Could we append these variables with `_bulk`? Or would that be too many code changes?

File: Source/Initialization/InjectorMomentum.H (hunk 1 above)
[the remark this answers is not in the document]
Comment (eloisejyang) [reply Source/Initialization/InjectorMomentum.H]: Returning `gamma*beta` would be consistent with what `getMomentum` returns for `u` in the Juttner and Gaussian momentum distributions as well.

File: Source/Initialization/InjectorMomentum.H [hunk 3]
```diff
@@ -108,6 +125,18 @@ struct InjectorMomentumBoltzmann
...
     }
 
+    AMREX_GPU_HOST_DEVICE
+    amrex::XDim3
+    getBulkMomentum (amrex::Real x, amrex::Real y, amrex::Real z) const noexcept
+    {
+        using namespace amrex;
+        Real u[3];
+        for (int idim = 0; idim < 3; ++idim) u[idim] = 0.0_rt;
+        Real gamma = 1./sqrt(1+beta*beta);
```
Comment (MaxThevenet) [inline Source/Initialization/InjectorMomentum.H]:
```suggestion
        const Real gamma = 1./sqrt(1.+beta*beta);
```

File: Source/Initialization/InjectorMomentum.H [hunk 4]
```diff
@@ -182,6 +211,18 @@ struct InjectorMomentumJuttner
...
```
Comment (MaxThevenet) [inline Source/Initialization/InjectorMomentum.H]:
```suggestion
        const Real gamma = 1./sqrt(1.+beta*beta);
```

File: Source/Initialization/InjectorMomentum.H [hunk 5]
```diff
@@ -203,6 +244,13 @@ struct InjectorMomentumRadialExpansion
         return {x*u_over_r, y*u_over_r, z*u_over_r};
     }
 
+    AMREX_GPU_HOST_DEVICE
+    amrex::XDim3
+    getBulkMomentum (amrex::Real x, amrex::Real y, amrex::Real z) const noexcept
+    {
+        return {x*u_over_r, y*u_over_r, z*u_over_r};
```
Comment (MaxThevenet) [inline Source/Initialization/InjectorMomentum.H]: This is correct provided `u_over_r` is in `m**[-1]`, i.e., `u_over_r` is the normalized momentum `gamma*beta` divided by the physical position (in SI). As this is not written anywhere in the code or in the doc, could you add a little doctsring in the `InjectorMomentumRadialExpansion` constructor to explain that? Thanks.

File: Source/Particles/PhysicalParticleContainer.cpp [hunk 6]
```diff
@@ -697,13 +705,15 @@ PhysicalParticleContainer::AddPlasma (int lev, RealBox part_realbox)
...
+                const XDim3 u_bulk = inj_mom->getBulkMomentum(x, y, 0.); // No z0_lab dependency
                 // At this point u is the lab-frame momentum
                 // => Apply the above formula for z0_lab
-                Real gamma_lab = std::sqrt( 1.+(u.x*u.x+u.y*u.y+u.z*u.z) );
-                Real betaz_lab = u.z/(gamma_lab);
-                Real z0_lab = gamma_boost * ( z*(1-beta_boost*betaz_lab)
-                                              - PhysConst::c*t*(betaz_lab-beta_boost) );
+                Real gamma_lab_bulk = std::sqrt(1.+(u_bulk.x*u_bulk.x+u_bulk.y*u_bulk.y+u_bulk.z*u_bulk.z));
+                Real betaz_lab_bulk = u.z/(gamma_lab_bulk);
+                Real z0_lab = gamma_boost * ( z*(1-beta_boost*betaz_lab_bulk)
```
Comment (MaxThevenet) [inline Source/Particles/PhysicalParticleContainer.cpp]:
```suggestion
                const Real gamma_lab_bulk = std::sqrt(1.+(u_bulk.x*u_bulk.x+u_bulk.y*u_bulk.y+u_bulk.z*u_bulk.z));
                const Real betaz_lab_bulk = u.z/(gamma_lab_bulk);
                const Real z0_lab = gamma_boost * ( z*(1.-beta_boost*betaz_lab_bulk)
```

File: Source/Particles/PhysicalParticleContainer.cpp [hunk 7]
```diff
@@ -672,12 +672,20 @@ PhysicalParticleContainer::AddPlasma (int lev, RealBox part_realbox)
...
                 // the next generated particle.
-                if (!inj_pos->insideBounds(xb, yb, z)) {
+
+                // include ballistic correction for plasma species with bulk motion
+                const XDim3 u_bulk = inj_mom->getBulkMomentum(x, y, z);
+                const Real gamma_bulk = std::sqrt(1.+(u_bulk.x*u_bulk.x+u_bulk.y*u_bulk.y+u_bulk.z*u_bulk.z));
+                const Real betaz_bulk = u.z/gamma_bulk;
+                const Real z0 = z - PhysConst::c*t*betaz_bulk;
+
+                if (!inj_pos->insideBounds(xb, yb, z0)) {
```
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.cpp:682]: this was z from the boosted frame before and is now z from the lab?

Comment (ax3l) [reply Source/Particles/PhysicalParticleContainer.cpp:682]: ah no, that's lab frame.

**Pull request merged**